text/3723-static-function-argument-unpacking.md

```diff
@@ -0,0 +1,882 @@
+- The struct fields need to be visible at call-site, e.g. `pub` or `pub(crate)`.
+- Attempting to unpack a struct with named fields, where the number and types of fields match, but the names are different, is rejected.
+    - Technically, under certain circumstances, it would be possible to emit syntactically correct code from the sugar, but the motivation is ambiguous. Therefore, it's better to leave it up to the developer to decide what is it that they want to accomplish.
+    - Also, it's difficult to specify what would happen when there are multiple arguments of the same type: What should the order be when the names don't match? What would happen if one of the struct's fields was renamed into one of the parameter names?
+
+However, several unresolved questions when unpacking structs would need to be considered as well:
+- What to do when unpacking structs with named fields into macro invocation's arguments?
+- How does unpacking more than one struct work?
+- How does unpacking structs combine with passing conventional arguments?
+- How does unpacking structs combine with unpacking tuples, tuple structs, and fixed-size arrays?
+- How to reconcile trait methods having differently named parameters?
+
+## Fallible Runtime Unpacking of Dynamic Collections
+
+The scope of argument unpacking could be expanded to dynamic contexts as well. Runtime unpacking of `dyn Trait` trait objects, slices, `Vec`s, `HashMap`s, iterators in general etc. would be fallible, since the existence of a correct number, order, typing and naming of items to match the parameters can't be guaranteed at compile time. A syntax such as `...expr?` or `...?expr` could be considered to improve ergonomics of argument passing for those cases as well, but that would definitely merit a separate RFC.
+
+Possibly, this would involve an stdlib trait, e.g. `TryArgUnpack`, whose implementation the language would use to get the arguments. This would enable unpacking custom collections as well.
+
+Interactions with or relation to the [`std::ops::Index`](https://doc.rust-lang.org/std/ops/trait.Index.html) and [`std::ops::IndexMut`](https://doc.rust-lang.org/std/ops/trait.IndexMut.html) traits would need to be mapped. Crucially, argument unpacking relies on a deterministic order for element access, which arbitrary implementations of `Index` cannot guarantee, suggesting of possible necessity of a deeper relation with `Iterator`.
+
+Infinite iterators, e.g. `std::iter::repeat`, would possibly have unwanted consequences when unpacked into variadic functions accepting infinite arguments.
+
+## Syntax Change for Functional Record Updates
+
+Adopting the `...expr` syntax for argument unpacking means that it is now part of the general "take stuff from here and expand it" family of syntactic sugar. As Rust already uses the `..` syntax for *Functional Record Updates*, changing that to use an ellipsis instead would be congruent.
```

Comment:
It might also be worthwhile to switch from `..` to `...` in [rest patterns](https://doc.rust-lang.org/reference/patterns.html#rest-patterns).

Reply:
that would be nice since it'd let you write the pattern `[a, ...b, c]` instead of having to write `[a, b @ .., c]` since just `..b` is a range pattern.